pkg/linter/lintFilterRuleSplitSubnet.go

```diff
@@ -0,0 +1,135 @@
+/*
+Copyright 2023- IBM Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package linter
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/np-guard/models/pkg/ipblock"
+	"github.com/np-guard/vpc-network-config-analyzer/pkg/vpcmodel"
+)
+
+const splitRuleSubnetName = "rules-splitting-subnets"
+
+// filterRuleSplitSubnet: rules of filters that are inconsistent w.r.t. subnets.
+type filterRuleSplitSubnet struct {
+	basicLinter
+}
+
+// a rule with the list of subnets it splits
+type splitRuleSubnet struct {
+	vpcName      string
+	rule         vpcmodel.RuleOfFilter
+	splitSubnets []vpcmodel.Subnet
+}
+
+// /////////////////////////////////////////////////////////
+// lint interface implementation for filterRuleSplitSubnet
+// ////////////////////////////////////////////////////////
+func (lint *filterRuleSplitSubnet) lintName() string {
+	return splitRuleSubnetName
+}
+
+func (lint *filterRuleSplitSubnet) lintDescription() string {
+	return "Firewall rules implying different connectivity for different endpoints within a subnet"
+}
+
+func (lint *filterRuleSplitSubnet) check() error {
+	for _, config := range lint.configs {
+		if config.IsMultipleVPCsConfig {
+			continue // no use in executing lint on dummy vpcs
+		}
+		for _, layer := range vpcmodel.FilterLayers {
+			filterLayer := config.GetFilterTrafficResourceOfKind(layer)
+			rules, err := filterLayer.GetRules()
+			if err != nil {
+				return err
+			}
+			for _, rule := range rules {
```

Comment:
if a rule in acl is splitting a subnet, but it has no effect on connectivity (for example if that nacl is not attached to the relevant subnet), it is also going to be reported?

Reply:
Yes, it is reported.
The question if it report always as now or to impose more filters is not trivial. Whether we report on potential problems or real problems.
1. I would surely not take consideration as redundant rule. We will have a different lint for Redundant rules.
2. `NACL` wise - should we consider only `relevant subnets` (which of course has to be defined)? Similar question can also be asked about SGs.  

I tend to think that we should keep it as it is now, for a few reasons:
1. We report best practices; it is up to the user to decide if its a problem or not. And breaking a subnet does not seem best practice.
1. Keep it simple and easy to define and understand. 

At the end of the day we should have many lints, each can be enabled or disabled. So perhaps we can have another, stricter, lint that reports only rules that results in different connectivity in the same subnet.

Reply:
I tend to agree with @ShiriMoran that linting should be simplistic and represent best practice even if it has no effect. Lots of examples like that in code linters.